### PR TITLE
add dycore test cases

### PR DESCRIPTION
**Pull request by szy21**
# Description

Add dycore test cases to one folder: experiments/TestCase/. Supersedes #1521 

- Move baroclinic_wave.jl from experiments/AtmosGCM/ to experiments/TestCase/ and make some minor changes
- Add risibingbubble.jl (but keep the risingbullel.jl in the tutorials/) and add a moist option
- Add solid_body_rotation.jl

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [x] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by szy21** [comment]
@akshaysridhar @jiahe23 Could you please take a look and make sure the dry rising bubble and dry baroclinic wave test cases I have here are useful test cases for the numerics? (I would like to add a moist rising bubble later.) Thanks!

**Comment by kpamnany** [comment]
Is this baroclinic wave different from the one in `experiments/AtmosGCM`? And is this dry rising bubble different from the one in `tutorials/Atmos`? There's also a baroclinic wave in the GCMDriver. Do we really want so many versions of this?

Shouldn't there be an `@test` or some sort of validation in the baroclinic wave experiment if this is for validation?

Perhaps we should meet to discuss?

**Comment by szy21** [comment]
> Is this baroclinic wave different from the one in `experiments/AtmosGCM`? And is this dry rising bubble different from the 
one in `tutorials/Atmos`? There's also a baroclinic wave in the GCMDriver. Do we really want so many versions of this?

They are the same. The reason I reorganize the drivers is we would like to have a faster turnaround for numerics development. I was thinking about a folder containing all the experiments that one should test for changes in numerics and would like to keep them as simple as possible. That's also the reason why I didn't use the GCMDriver. Having said that, I would be happy to remove the `baroclinic-wave` in` experiments/AtmosGCM`. `risingbubble` is a bit more complicated because it's a tutorial.

> Shouldn't there be an `@test` or some sort of validation in the baroclinic wave experiment if this is for validation?
> Perhaps we should meet to discuss?

These "tests" are a bit hard to write as real tests because they require someone to visualize the results and see whether they make sense. I would be happy to talk!

**Comment by szy21** [comment]
Following the discussion with @kpamnany I am going to add these experiments to CI and remove `baroclinic-wave` in `experiments/AtmosGCM/`. Kiran, could you please add an argument for `timeend`?

**Comment by szy21** [comment]
This is ready for review. @jiahe23 @kpamnany could you please take a look at whether the drivers look good to you? @jakebolewski could you please take a look at whether the modification of pipeline.yml looks good? Thanks!

**Comment by szy21** [comment]
bors try

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -897,8 +897,8 @@ steps:
       slurm_ntasks: 1
       slurm_gres: "gpu:1"
 
-  - label: "gpu_risingbubble"
-    key: "gpu_risingbubble"
+  - label: "gpu_risingbubble_tutorial"
+    key: "gpu_risingbubble_tutorial"
     command:
       - "mpiexec julia --color=yes --project tutorials/Atmos/risingbubble.jl "
```
Comment (kpamnany) [inline .buildkite/pipeline.yml:903]: Do we still need to run the tutorial version for CI?

File: experiments/TestCase/risingbubble.jl [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    θ = θ_ref + Δθ                                      # potential temperature
+    q_pt = PhasePartition(Δq_tot)
+    R_m = gas_constant_air(bl.param_set, q_pt)
+    _cp_m = cp_m(bl.param_set, q_pt)
+    _cv_m = cv_m(bl.param_set, q_pt)
+    π_exner = FT(1) - _grav / (_cp_m * θ) * z           # exner pressure
+    ρ = p0 / (R_gas * θ) * (π_exner)^(_cv_m / R_m)      # density
+    T = θ * π_exner
+
+    if bl.moisture isa EquilMoist
```
Comment (kpamnany) [inline experiments/TestCase/risingbubble.jl:66]: Should this work even with `NonEquilMoist`? If so, maybe testing for `DryModel` is better.

**Review by kpamnany — APPROVED** [review]
LGTM. But I think these should move to `test/validation/` instead of `experiments/TestCase`?

**Review by jiahe23 — APPROVED** [review]

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (szy21) [reply .buildkite/pipeline.yml:903]: I'm fine with removing it. @akshaysridhar ?

**Comment by szy21** [comment]
> LGTM. But I think these should move to `test/validation/` instead of `experiments/TestCase`?

I was thinking about it. The problem is for some tests (e.g., baroclinic wave), it would be hard to have @test or some sort of validation without looking at the results by eye. In that case, should we put these in test/?

File: experiments/TestCase/risingbubble.jl (hunk 2 above)
Comment (szy21) [reply experiments/TestCase/risingbubble.jl:66]: I'm not sure whether this works with `NonEquilMoist`, but I think it's better to test for `DryModel`. I will change it.

File: .buildkite/pipeline.yml [hunk 3]
```diff
@@ -907,26 +907,66 @@ steps:
...
+    agents:
+      config: gpu
+      queue: central
+      slurm_ntasks: 1
+      slurm_gres: "gpu:1"
+
+  - label: "gpu_moist_risingbubble"
+    key: "gpu_moist_risingbubble"
+    command:
+            - "mpiexec julia --color=yes --project experiments/TestCase/risingbubble.jl --with-moisture "
```
Comment (jakebolewski) [inline .buildkite/pipeline.yml:923]: just minor formatting here, YAML is indentation sensitive in a block so it's nice to keep it consistent

File: .buildkite/pipeline.yml [hunk 4]
```diff
@@ -907,26 +907,66 @@ steps:
...
     agents:
       config: gpu
       queue: central
       slurm_ntasks: 1
       slurm_gres: "gpu:1"
 
+  - label: "gpu_moist_baroclinic_wave"
+    key: "gpu_moist_baroclinic_wave"
+    command:
+            - "mpiexec julia --color=yes --project experiments/TestCase/baroclinic_wave.jl --with-moisture "
```
Comment (jakebolewski) [inline .buildkite/pipeline.yml:953]: and here

File: .buildkite/pipeline.yml (hunk 3 above)
Comment (szy21) [reply .buildkite/pipeline.yml]: Thanks!

File: .buildkite/pipeline.yml (hunk 1 above)
Comment (szy21) [reply .buildkite/pipeline.yml]: I removed this from CI.

**Comment by szy21** [comment]
bors try

**Comment by jkozdon** [comment]
> Build failed:
> 
> * [buildkite/climatemachine-ci](https://buildkite.com/clima/climatemachine-ci/builds/488)

@blallen Seem to be an ocean test: https://buildkite.com/clima/climatemachine-ci/builds/488#cc5633bd-09a5-4805-a819-d3c5e5e2d9c0

**Comment by jkozdon** [comment]
> Build failed:
> 
> * [buildkite/climatemachine-ci](https://buildkite.com/clima/climatemachine-ci/builds/489)

Another ocean failure: https://buildkite.com/clima/climatemachine-ci/builds/489#a1f3aee0-fb90-47c4-81b4-0928f1d8e337

**Comment by charleskawczynski** [comment]
I see `cp` and stuff going on inside the build. I think we should change
`joinpath("test", "Land", "Model", "Artifacts.toml"),`
to
`joinpath(@__DIR__, "Artifacts.toml"),`.

**Comment by szy21** [comment]
> Should we change
> `joinpath("test", "Land", "Model", "Artifacts.toml"),`
> to
> `joinpath(@__DIR__, "Artifacts.toml"),`?

Should we change it in this PR? I didn't touch these test files.

**Comment by blallen** [comment]
> > Build failed:
> > 
> > * [buildkite/climatemachine-ci](https://buildkite.com/clima/climatemachine-ci/builds/489)
> 
> Another ocean failure: https://buildkite.com/clima/climatemachine-ci/builds/489#a1f3aee0-fb90-47c4-81b4-0928f1d8e337

Both of these look like real errors from the logs, but I can't see anything in this PR that would cause that.

**Pull request merged**